src/Http/Routing/src/RouteEndpointBuilder.cs

```diff
@@ -39,15 +44,39 @@ public sealed class RouteEndpointBuilder : EndpointBuilder
     }
 
     /// <inheritdoc />
+    [UnconditionalSuppressMessage("Trimmer", "IL2026",
+        Justification = "We surface a RequireUnreferencedCode in AddRouteHandlerFilter which is required to call unreferenced code here. The trimmer is unable to infer this.")]
     public override Endpoint Build()
     {
         if (RequestDelegate is null)
         {
             throw new InvalidOperationException($"{nameof(RequestDelegate)} must be specified to construct a {nameof(RouteEndpoint)}.");
         }
 
+        var requestDelegate = RequestDelegate;
+
+        // Only replace the RequestDelegate if filters have been applied to this builder and they were not already handled by RouteEndpointDataSource.
+        // This affects other data sources like DefaultEndpointDataSource (this is people manually newing up a data source with a list of Endpoints),
+        // ModelEndpointDataSource (Map(RoutePattern, RequestDelegate) and by extension MapHub, MapHealChecks, etc...),
+        // ActionEndpointDataSourceBase (MapControllers, MapRozorPages, etc...) and people with custom data sources or otherwise manually building endpoints
+        // using this type. At the moment this class is sealed, so at the moment we do not need to concern ourselves with what derived types may be doing.
+        if (RouteHandlerFilterFactories is { Count: > 0 })
+        {
+            // Even with filters applied, RDF.Create() will return back the exact same RequestDelegate instance we pass in if filters decide not to modify the
+            // invocation pipeline. We're just passing in a RequestDelegate so none of the fancy options pertaining to how the Delegate parameters are handled
+            // do not matter.
+            RequestDelegateFactoryOptions rdfOptions = new()
+            {
+                RouteHandlerFilterFactories = RouteHandlerFilterFactories,
+                EndpointMetadata = Metadata,
+            };
+
+            // We ignore the returned EndpointMetadata has been already populated since we passed in non-null EndpointMetadata.
+            requestDelegate = RequestDelegateFactory.Create(requestDelegate, rdfOptions).RequestDelegate;
+        }
```

Comment:
While I think this is cool and I now understand it more, can we do this in a later change after we discuss the implications? If we enable this it might make sense to rename `RouteHandlerFilterInvocationContext` as it makes things more general purpose. I also am not sure about the coupling to the request delegate factory. That seems unnecessary. I would just rewrite the logic here since there's no need for codegen.